File: performance_test/bench_louvain.py
#!/usr/bin/env python3
"""Louvain: axolotl-rs vs NetworkX vs iGraph vs Networkit"""

import time, os, sys
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from bench_comparison import load_edges

# Build wheel first
os.chdir("/tmp/axolotl-rs/prototype-rust")
os.system("/Users/sai/.workbuddy/binaries/python/versions/3.13.12/bin/maturin build --release --features python-bindings 2>/dev/null")
os.system("~/.workbuddy/venvs/lobster-memory/bin/pip install --force-reinstall target/wheels/axolotl_rs-0.1.0-cp313-cp313-macosx_11_0_arm64.whl 2>/dev/null")

os.chdir("/Users/sai/WorkBuddy/点子/performance_test")

def bench_axolotl(name, edges, n_v):
    import axolotl_rs
    g = axolotl_rs.AxolotlGraph.new()
    for i in range(n_v): g.add_vertex(i)
    for u, v in edges:
        g.add_edge(u, v)
        g.add_edge(v, u)
    t0 = time.perf_counter()
    # Louvain not exposed via Python binding — use subprocess to call Rust binary
    elapsed = (time.perf_counter() - t0) * 1000
    return elapsed, 0

def bench_networkx(name, edges, n_v):
    import networkx as nx
    from networkx.algorithms.community import louvain_communities
    g = nx.Graph()
    g.add_nodes_from(range(n_v))
    g.add_edges_from(edges)
    t0 = time.perf_counter()
    comms = louvain_communities(g)
    elapsed = (time.perf_counter() - t0) * 1000
    return elapsed, len(comms)

def bench_igraph(name, edges, n_v):
    import igraph
    g = igraph.Graph(directed=False)
    g.add_vertices(n_v)
    g.add_edges(edges)
    t0 = time.perf_counter()
    comms = g.community_multilevel()
    elapsed = (time.perf_counter() - t0) * 1000
    return elapsed, len(comms)

def bench_networkit(name, edges, n_v):
    import networkit as nk
    g = nk.Graph(n_v, directed=False)
    for u, v in edges:
        g.addEdge(u, v)
    t0 = time.perf_counter()
    cd = nk.community.PLM(g)
    cd.run()
    elapsed = (time.perf_counter() - t0) * 1000
    return elapsed, cd.numberOfSubsets()

def main():
    # First get axolotl-rs numbers from Rust benchmark
    import subprocess
    os.chdir("/tmp/axolotl-rs/prototype-rust")
    result = subprocess.run(
        ["cargo", "run", "--release", "--example", "bench_louvain"],
        capture_output=True, text=True, timeout=300
    )
    ax_data = {}
    for line in result.stdout.split('\n'):
        if '|' in line and 'ms' in line and 'Dataset' not in line:
            parts = [p.strip() for p in line.split('|') if p.strip()]
            if len(parts) >= 5 and parts[0] != '---------':
                name = parts[0]
                time_str = parts[3].replace('ms', '')
                comms_str = parts[4]
                try:
                    ax_data[name] = (float(time_str), int(comms_str))
                except: pass

    os.chdir("/Users/sai/WorkBuddy/点子/performance_test")

    datasets = [
        ("soc-Epinions1", "datasets/soc_epinions1_100K.edgelist"),
        ("com-DBLP", "datasets/com_dblp_100K.edgelist"),
        ("RMAT scale 20", "datasets/rmat_1M_100K.edgelist"),
    ]

    print("# Louvain Community Detection — Library Comparison\n")
    print(f"**Hardware**: Apple M4\n")
    print("| Dataset | V | axolotl-rs Louvain | NetworkX | iGraph (C) | Networkit (C++) |")
    print("|---------|:---:|:---:|:---:|:---:|:---:|")

    for name, path in datasets:
        if not os.path.exists(path): continue
        edges, n_v = load_edges(path)
        v_str = f"{n_v//1000}K" if n_v < 1_000_000 else f"{n_v/1e6:.1f}M"

        t_nx, c_nx = bench_networkx(name, edges, n_v)
        t_ig, c_ig = bench_igraph(name, edges, n_v)
        t_nk, c_nk = bench_networkit(name, edges, n_v)

        ax_str = "—"
        if name in ax_data:
            t_ax, c_ax = ax_data[name]
            ax_str = f"{t_ax:.0f}ms ({c_ax})"

        print(f"| {name} | {v_str} | {ax_str} | {t_nx:.0f}ms ({c_nx}) | {t_ig:.0f}ms ({c_ig}) | {t_nk:.0f}ms ({c_nk}) |")

if __name__ == "__main__":
    main()

File: prototype-rust/examples/bench_louvain.rs
/// examples/bench_louvain.rs — Louvain (EdgeBlock) benchmark

use std::collections::HashMap;
use std::fs;
use std::io::{BufRead, BufReader};
use std::path::Path;
use std::time::Instant;

use axolotl_rs::gpu_edge_block::GPUEdgeBlockGraph;
use axolotl_rs::louvain::louvain_communities;

fn load_edgelist(path: &str) -> (Vec<(u64, u64)>, usize) {
    let f = fs::File::open(path).expect("open");
    let mut edges = Vec::new();
    let mut max_v = 0u64;
    for line in BufReader::new(f).lines() {
        let line = line.unwrap();
        if line.is_empty() || line.starts_with('#') || line.starts_with('%') { continue; }
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 {
            if let (Ok(u), Ok(v)) = (parts[0].parse::<u64>(), parts[1].parse::<u64>()) {
                if u != v { edges.push((u, v)); if u > max_v { max_v = u; } if v > max_v { max_v = v; } }
            }
        }
    }
    (edges, (max_v + 1) as usize)
}

fn format_num(n: usize) -> String {
    if n >= 1_000_000 { format!("{:.1}M", n as f64 / 1_000_000.0) }
    else if n >= 1_000 { format!("{}K", n / 1_000) }
    else { n.to_string() }
}

fn main() {
    let datasets: Vec<(&str, &str)> = vec![
        ("soc-Epinions1", "performance_test/datasets/soc_epinions1_100K.edgelist"),
        ("com-DBLP", "performance_test/datasets/com_dblp_100K.edgelist"),
        ("web-Google", "performance_test/datasets/web_google_100K.edgelist"),
        ("RMAT scale 20", "performance_test/datasets/rmat_1M_100K.edgelist"),
    ];

    println!("# Louvain Community Detection (axolotl-rs EdgeBlock)\n");
    println!("| Dataset | V | E | Time | Communities | Passes |");
    println!("|---------|:---:|:---:|:---:|:---:|:---:|");

    for (name, path) in &datasets {
        if !Path::new(path).exists() { continue; }
        let (edges, n_v) = load_edgelist(path);

        let mut eb = GPUEdgeBlockGraph::new();
        for i in 0..n_v as u64 { eb.add_vertex(i, HashMap::new()); }
        for &(u, v) in &edges { eb.add_edge(u, v, 1.0); eb.add_edge(v, u, 1.0); }

        let t0 = Instant::now();
        let (result, passes) = louvain_communities(&eb);
        let elapsed = t0.elapsed().as_secs_f64() * 1000.0;
        let n_comms = result.iter().max().map(|&x| x + 1).unwrap_or(0);

        println!("| {} | {} | {} | {:.0}ms | {} | {} |",
            name, format_num(n_v), format_num(edges.len()), elapsed, n_comms, passes);
    }
}
